### PR TITLE
Add Fadt structure

### LINKED ISSUE #2: Parse the FADT
One of the first tables we should be able to parse is the FADT (and it's one of the easier ones), which tells us about register blocks for power management.

We should:
1) Create a struct to represent the structure in Rust. It needs to be `#[repr(C, packed)]` and needs to match the layout the hardware has constructed. Either look at the relevant section of the [spec](http://www.uefi.org/sites/default/files/resources/ACPI%206_2_A_Sept29.pdf) or [this helpful page](https://wiki.osdev.org/FADT) to start
2) Create a function `parse_fadt(mapping : PhysicalMapping<Fadt>)`. This should validate and then parse the table.
3) Work out which fields to use for the FACS and DSDT addresses, once again there's helpful info [here](https://wiki.osdev.org/FADT)
4) Dispatch to `parse_fadt` in `dispatch_sdt`

There might be more info we care about from the FADT in the future, but this is probably a good start.

### PR DESCRIPTION
**Pull request by Sh4d1**
Start to fix #2 

Signed-off-by: Patrik Cyvoct <patrik@ptrk.io>

### CODE REVIEW

**Comment by Sh4d1** [comment]
@IsaacWoods I just added the Fadt struct, good for you? Still thinking of how to implement tests

File: src/sdt.rs [hunk 1]
```diff
@@ -130,6 +131,17 @@ pub(crate) fn dispatch_sdt<H>(handler : &mut H, physical_address : usize) -> Res
          */
         match signature
         {
+            "FADT" => {
+                let fadt_mapping = handler.map_physical_region::<Fadt>(physical_address);
+                match parse_fadt(fadt_mapping) {
```
Comment (IsaacWoods) [inline src/sdt.rs]: I think it'd be better to lend the mapping to `parse_fadt` and then unmap it here, so the mapping and unmapping logic is in one place

File: src/sdt.rs [hunk 2]
```diff
@@ -52,7 +53,7 @@ impl SdtHeader
         }
 
         // Check that the lowest byte is 0
-        if sum & 0b1111_1111 != 0
+        if sum % 0b1111_1111 != 0
```
Comment (IsaacWoods) [inline src/sdt.rs]: I'm probably just having a brain-fart, but why this change?

File: src/fadt.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    x_pm_timer_block: GenericAddress,
+    x_gpe0_block: GenericAddress,
+    x_gpe1_block: GenericAddress,
+    sleep_control_reg: GenericAddress,
+    sleep_status_reg: GenericAddress,
+    hypervisor_vendor_id: u64,
+}
+
+impl Fadt {
+    pub fn validate(&self) -> Result<(), AcpiError> {
```
Comment (IsaacWoods) [inline src/fadt.rs]: `SdtHeader::validate` was updated in [this commit](https://github.com/rust-osdev/acpi/commit/f578070927d97bdf891c1c80bcdd51e2ebab4b3f), so this function could be replaced with `self.header.validate(b"FACP")` at the call-site. You'll probably need to rebase anyways, so might as well update this

File: src/sdt.rs (hunk 2 above)
Comment (Sh4d1) [reply src/sdt.rs]: It's not clear for FADT in the docs but as for RSDT: All bytes of the table summed must be equal to 0 (mod 0xFF). And since sum is usize it will grow beyond 0xFF and if you do a bitwise and on it, it will not be the same result as if you would use the mod. Another solution would be to use a u8 but not sure what happens when you grow over b1111_1111. Not sure if I was clear

File: src/fadt.rs (hunk 3 above)
Comment (Sh4d1) [reply src/fadt.rs]: After a painfull rebase and merge (I think I fucked up somewhere haha), it is changed :smile:

Comment (Sh4d1) [reply src/fadt.rs]: Also changed `==` to `!=` in the validate signature :p

File: src/sdt.rs (hunk 1 above)
Comment (Sh4d1) [reply src/sdt.rs]: Not sure I understood what you meant

File: src/sdt.rs (hunk 2 above)
Comment (IsaacWoods) [reply src/sdt.rs]: Huh, did you see a section in the spec about mod? I was basing my implementation on [here](https://wiki.osdev.org/RSDP) where it says "you add up every byte in the structure and make sure the lowest byte of the result is equal to zero.". Looking around, [Haiku also appears to use `&`](https://github.com/haiku/haiku/blob/master/src/system/boot/platform/efi/acpi.cpp#L50).

To be honest tho, I'm not sure I understand the difference enough to be sure. I'll have a think

Comment (Sh4d1) [reply src/sdt.rs]: Okay, so maybe RSDT and RSDP do not have the same way of checking. For RSDT see [here](https://wiki.osdev.org/RSDT#Checksum).

Comment (Sh4d1) [reply src/sdt.rs]: And regarding the implementation of Haiku, they use a `unsigned char` which is only 8 bits, so when the reach the max value it goes back to 0

File: src/fadt.rs (hunk 3 above)
Comment (IsaacWoods) [reply src/fadt.rs]: Oh wow; that's why we needs tests I guess :laughing:. And looks like you've included all of the changes again instead of rebasing, but it'll apparently still auto-merge, so I'm cool with that. Git is such as pain sometimes.

File: src/sdt.rs (hunk 2 above)
Comment (Sh4d1) [reply src/sdt.rs]: So yeah I think we should also use `%` in the RSDP

File: src/fadt.rs (hunk 3 above)
Comment (Sh4d1) [reply src/fadt.rs]: Yeah I forgot to squash, will do it later so the log history won't be too ugly :p

File: src/sdt.rs (hunk 1 above)
Comment (IsaacWoods) [reply src/sdt.rs]: So take a `&PhysicalMapping<Fadt>` instead in `parse_fadt`, change this match to `parse_fadt(&fadt_mapping)` and then do `handler.unmap_physical_region(fadt_mapping)` after the match

File: src/sdt.rs [hunk 4]
```diff
@@ -117,6 +118,14 @@ where
          * and length, and then the dispatched to the correct function to actually parse the table.
          */
         match signature {
+            "FADT" => {
+                let fadt_mapping = handler.map_physical_region::<Fadt>(physical_address);
+                match parse_fadt(&fadt_mapping) {
```
Comment (IsaacWoods) [inline src/sdt.rs]: This could just be `parse_fadt(&fadt_mapping)?;`

File: src/sdt.rs [hunk 5]
```diff
@@ -117,6 +118,14 @@ where
...
+                    Ok(()) => {}
+                    Err(err) => panic!("Error parsing FADT: {:#?}", err),
+                }
+                handler.unmap_physical_region(fadt_mapping);
+            }
```
Comment (IsaacWoods) [inline src/sdt.rs]: I think you'll need a comma here after the match arm

Comment (Sh4d1) [reply src/sdt.rs]: Not necesseraly, but required when using one line statements. And rustfmt removes it so .. :smile:

File: src/sdt.rs (hunk 4 above)
Comment (Sh4d1) [reply src/sdt.rs]: Yep :wink:

File: src/sdt.rs (hunk 5 above)
Comment (IsaacWoods) [reply src/sdt.rs]: Huh, fair enough

**Comment by IsaacWoods** [comment]
Okay, cool! I'm happy with everything here except maybe the checksum thing, which we can deal with at a later date. When you're happy, do a final squash and then I'll merge!

**Comment by Sh4d1** [comment]
@IsaacWoods done! Regarding the checksum, I think it's the same way for both.
Basicall either 
 * use an u8 and `&` or `%`
 * use a usize and `%`

**Comment by IsaacWoods** [comment]
Thanks so much! And yeah I think you're right, I'll try it out for the other structures

bors r+

File: src/sdt.rs (hunk 2 above)
Comment (IsaacWoods) [reply src/sdt.rs]: Yup, that sounds right. I don't think using the equivalent in Rust (using `u8` and letting it go back to 0) would even work because it would just panic on the overflow, so let's move to using `%`
